tensorflow/core/lib/io/buffered_random_access_file.cc

```diff
@@ -0,0 +1,124 @@
+#include "tensorflow/core/lib/io/buffered_random_access_file.h"
+
+namespace tensorflow {
+namespace io {
+
+BufferedRandomAccessFile::BufferedRandomAccessFile(
+    std::unique_ptr<RandomAccessFile>&& other_file, size_t buf_size)
+    : BufferedRandomAccessFile(other_file.release(), buf_size, true) {}
+
+BufferedRandomAccessFile::BufferedRandomAccessFile(RandomAccessFile* other_file,
+                                                   size_t buf_size,
+                                                   bool own_file)
+    : buf_size_(buf_size),
+      buf_(new char[buf_size_]),
+      pos_(0),
+      limit_(0),
+      file_pos_(0),
+      own_file_(own_file),
+      other_file_(other_file) {}
+
+BufferedRandomAccessFile::~BufferedRandomAccessFile() {
+  mutex_lock lock(mu_);
+  if (own_file_) {
+    delete other_file_;
+  }
+  delete[] buf_;
+
+  buf_ = nullptr;
+  other_file_ = nullptr;
+}
+
+Status BufferedRandomAccessFile::Read(uint64 offset, size_t n,
+                                      StringPiece* result,
+                                      char* scratch) const {
+  mutex_lock lock(mu_);
+  Seek(offset);
+
+  Status status;
+  size_t total_read = 0;
+  while (total_read < n) {
+    if (pos_ == limit_) {
+      if (n - total_read >= buf_size_) {
+        // Unnecessary to read into buffer and then copy to scratch when buffer
+        // is empty and byte to read is large enough.
+        StringPiece tmp_result;
+        status = other_file_->Read(offset + total_read, n - total_read,
+                                   &tmp_result, scratch + total_read);
+        if (tmp_result.length() > 0 &&
+            tmp_result.data() != scratch + total_read) {
```

Comment:
I don't think I completely understand this test here. Can you please add a comment?

Reply:
Comment added.